src/net/sourceforge/kolmafia/request/RelayRequest.java

```diff
@@ -3500,6 +3514,156 @@ public boolean pollForCompletion(String type) {
     return true;
   }
 
+  private void jsonError(String message) {
+    this.statusLine = "HTTP/1.1 400 Bad Request";
+    this.contentType = "application/json";
+    this.responseCode = 400;
+    this.responseText = JSON.toJSONString(Map.of("error", message));
+  }
+
+  private String underscoreName(String name) {
+    return name.replaceAll("[A-Z]", "_$0").toLowerCase();
+  }
+
+  private Value transformApiArgument(Object thing) {
+    if (thing instanceof JSONObject obj) {
+      var objectTypeObject = obj.get("objectType");
+      Type dataType;
+      if (objectTypeObject instanceof String objectType
+          && ((dataType = DataTypes.enumeratedTypes.find(objectType.toLowerCase())) != null)) {
+        var identifierStringObject = obj.get("identifierString");
+        var identifierNumberObject = obj.get("identifierNumber");
+        if (identifierStringObject instanceof String identifierString) {
+          return dataType.parseValue(identifierString, true);
+        } else if (identifierNumberObject instanceof Integer identifierNumber) {
+          return dataType.makeValue(identifierNumber, true);
+        }
+      }
+
+      return JSONValueConverter.fromJSON(obj);
+    } else return JSONValueConverter.fromJSON(thing);
+  }
+
+  /**
+   * Handle JSON API request. Format is e.g. { properties: ["kingLiberated"], functions: [{ name:
```

Comment:
This isn't comprehensive enough for me to understand how it's supposed to work.

Please make an actual spec instead of an "e.g." (TypeScript-style is fine, pseudocode is also fine, doesn't need to be overly formal but should be exhaustive).

There are some misspellings ("myTunrcount", "can be request") that I presume will be fixed with a rewrite.

I don't understand what "Enumerated types will be returned in the same format." is supposed to mean. Looking at the tests, it doesn't mean "the same format you passed in", which is what I thought this meant.

You also need to specify how non-enumerated types are returned -- with only the identifierString?

Looking at the tests it seems "objectType" is always capitalized when returned, so these examples are also wrong.